### ISSUE TITLE
dkim signature not rfc2822 strict

### ISSUE THREAD

**Issue by joaoreis81** [postalserver/postal#339]
DKIM header isn't folded with 78 chars long (as it *should*), which may be the cause of bogus results in some dkim verification.

**Comment by igerzog** [comment]
Taking the exact workaround to match different checking filters.

**Comment by catphish** [comment]
Thanks, I'll look into making sure these are compliant.

**Comment by mattmartincouk** [comment]
We've been testing DKIM via native mail. For hotmail, in our testing none of our mail landed in the inbox but when we sent via Postal, it did.

Content was similar on all messages pairs to give a fair test.

**Comment by mattmartincouk** [comment]
We're now getting resports that some mail to Gmail addresses is being classed as spam due to invalid DKIM. Quite serious as we're getting poor deliverability now. Is there any chance this one could be fixed due to the implications it has on the primary function of mail deliverability?

**Comment by catphish** [comment]
Yes, this does seem important, sorry for forgetting to look. Will try to get onto it asap.

**Comment by willpower232** [comment]
We haven't noticed any problems with deliverability but referencing PHPMailer, I can see they break down the DKIM header. I have made a modification to our test environment which doesn't appear to have any negative feedback to accomplish this.

I have changed https://github.com/atech/postal/blob/6f29bc8d3f6f197d142724a7baee7376ff5fe8aa/lib/postal/message_db/message.rb#L403 to read as follows

```ruby
headers << dkim.dkim_header.scan(/.{1,76}/).join("\r\n")
```

I restarted Postal to make this change take affect. Also I noted that PHPMailer said to split on 76 characters so that `\r\n` made up the 77th and 78th characters.

I'm happy to make a pull request if this was the correct modification to make but I'd hope you can test to see if this improves your deliverability in the meantime.

**Comment by joaoreis81** [comment]
Just for information, I had problems signing messages with with using \r\n with a sendmail milter that a I wrote, had to change it to \n. Got some "Message containing bare LF's." errors trying to deliber to some MXs.

**Comment by willpower232** [comment]
@joaoreis81 I understand that `\r\n` is mandated by SMTP so if you remove them, you should restore them before sending the message on.

I'm not sure if my modification is correct, it might need some white space at the start of each split line, i.e.  `.scan(/.{1,75}/).join("\r\n\t")` but I can't see this in PHPMailer. I'm hoping @mattmartincouk can clarify what is needed to alleviate his situation.

**Comment by mattmartincouk** [comment]
The first code sample didn't seem to work at all.

The second did but still Gmail (and AOL) seem to spam messages as DKIM fails to validate and thus our DMARC policy says to spam.

Checking the databases for the DMARC SMTP response, shows that most mail is fine but some are not. Can't correlate anything but the same message not via postal isn't caught.

I can repeat the failure with the same message to GMail over and over so can be sure its not a momentary glitch.

**Comment by willpower232** [comment]
Are you able to determine whether the problem happens with just text or just HTML or if it requires both? 

Can you then share the problematic message so we can replicate?

**Comment by catphish** [comment]
I have looked into this and I have a few opinions:

Postal's DKIM signatures are indeed longer than the recommended 78 characters, but they are not longer than the maximum 998 characters. This should therefore not be the cause of the problem, but should definitely be fixed.

@willpower232's patch is close but in my tests, it has the potential to wrap the signature part-way through a value other than the signature, which I would suggest is not a good idea:

```
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=email.codebasehq.co
	m; s=postal-OpjvKU; t=1516101424; bh=mvuu3b9qEfdaZCSaVxLvRSqKbMB5YN4qxRCRw5
	FtvuE=; h=date:from:to:message-id:subject; b=eABPHB69nlL+jbelwqrMLufzGhm7KU
	ETzWvT4+2qOkGJ/rO9gvLpr012hvNzWmFqlmUu4Ukkj+779MloKLiP5HRHvZXRY24xpCrq+IL/i
	knvTCNBIjCX3hj77WyUO4qKTy7ANAFVHhHC4uKGUKoL7Y/b7Wi8Md3AaB9iQwdyhq0=
```

I would suggest that @mattmartincouk's issue is actually unrelated, as he says wrapping the header does not resolve the issue.

Note, the whitespace at the beginning of the wrapped lines is definitely required, else the next line is just considered a new header.

**Comment by catphish** [comment]
@mattmartincouk and/or @joaoreis81 please could you test this fix ebf3d7932b56fa49a6de933001868b1e31e38526

**Comment by catphish** [comment]
It's entirely possible that this **is** Matt's issue, since the initial fix may not have fixed it for him due to the bad wrapping part way through words. Here's an example of the new output:

```
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
	d=nutty.tk;
	s=postal-uIzkKZ; t=1516106068;
	bh=mvuu3b9qEfdaZCSaVxLvRSqKbMB5YN4qxRCRw5FtvuE=;
	h=date:from:to:message-id:subject;
	b=FgwgOGvAonYvjZLmIxScwEhw38Df5YBj0Qh4PgGhjsigq8QTh7EEA8hfGXooRDRjvyH+YOQc
	3u1X9J4NlJFis9d/SjCMW6KyTLBE0iI7ZYSJ0GeB5QDbUqDjNo0xkY48E2UTWa3rVjXDF86f
	Y6w3dlH/lxo0J+F1e/RwAAgRuKU=
```
